### DISCUSSION TITLE
How to unit test streaming RPCs

### DISCUSSION THREAD

**Discussion by jonnylangefeld** [connectrpc/connect-go#675] [Q&A]
I have a unit test file for every `service`. In the unit test I instantiate the server like
```
server = NewHello()
```

And then just call the functions directly. Check below for a working example

<details>
<summary>unfold for working unit test</summary>

```go
var _ = Describe("Hello", func() {
	var server *Hello

	BeforeEach(func() {
		server = NewHello()
	})

	Context("SayHello", func() {
		It("returns a response", func() {
			resp, err := server.SayHello(nil, &connect.Request[lifev1.SayHelloRequest]{
				Msg: &lifev1.SayHelloRequest{
					Name: "Test Name",
				},
			})
			Expect(err).NotTo(HaveOccurred())
			Expect(resp.Msg.Message).To(Equal("Hello Test Name"))
		})

		It("returns an error if the name is missing", func() {
			_, err := server.SayHello(nil, &connect.Request[lifev1.SayHelloRequest]{
				Msg: &lifev1.SayHelloRequest{
					Name: "",
				},
			})
			Expect(err).To(HaveOccurred())
			Expect(err.Error()).To(ContainSubstring("missing name"))
		})
	})

})
```

</details>

But now for a different service I have an rpc with a stream:

```go
func (s *MyService) WatchUnits(_ context.Context, in *connect.Request[lifev1.WatchUnitsRequest], stream *connect.ServerStream[lifev1.WatchUnitsResponse]) error {
	for update := range s.updates {
		if err := stream.Send(update); err != nil {
			return err
		}
	}
	return nil
}
```
(`s.updates` is a channel)

How can I test this?

This is my attempt, but it's not quite right yet:

```go
	Describe("WatchUnits", func() {
		It("receives messages on the stream", func() {
			stream := &connect.ServerStream[lifev1.WatchUnitsResponse]{}
			go func() {
				Expect(server.WatchUnits(nil, nil, stream)).NotTo(HaveOccurred())
			}()
			v := &lifev1.WatchUnitsResponse{}
			go func() {
				server.updates <- v
			}()
			Expect(stream).Should(Receive(v))
		})
	})
```

**Comment by emcfarlane** [comment]
Hey @jonnylangefeld , currently these constructors for the Stream types aren't exported. See the docs for the stream types: e.g. https://pkg.go.dev/connectrpc.com/connect#BidiStream
> doesn't currently have an exported constructor.

To test handlers the current recommended way is to use the [net/http/httptest](https://pkg.go.dev/net/http/httptest) package and create a server with a corresponding client to invoke the server. See the tests in the connect package as examples: https://github.com/connectrpc/connect-go/blob/297532d3e86d55f51a610c424f4b7a4fa363ab21/client_ext_test.go#L446
